vms/platformvm/state/state.go

```diff
@@ -121,11 +125,40 @@ type State interface {
 	// [vdrs].
 	ValidatorSet(subnetID ids.ID, vdrs validators.Set) error
 
-	GetValidatorWeightDiffs(height uint64, subnetID ids.ID) (map[ids.NodeID]*ValidatorWeightDiff, error)
-
-	// Returns a map of node ID --> BLS Public Key for all validators
-	// that left the Primary Network validator set.
-	GetValidatorPublicKeyDiffs(height uint64) (map[ids.NodeID]*bls.PublicKey, error)
+	// ApplyValidatorWeightDiffs iterates from [startHeight] towards the genesis
+	// block until it has applied all of the diffs through [endHeight]. Applying
+	// the diffs results in modifying [validators].
+	//
+	// Invariant: If attempting to generate the validator set for
+	// [endHeight - 1], [validators] should initially contain the validator
+	// weights for [startHeight].
+	//
+	// Note: Because this function iterates towards the genesis, [startHeight]
+	// should normally be greater than or equal to [endHeight].
+	ApplyValidatorWeightDiffs(
```

Comment:
nit: I'm surprised this isn't unified with `ApplyValidatorPublicKeyDiffs`...are they often used separately in practice?

Reply:
It's difficult to implement these as a combined function - because when generating a subnet validator set, we aren't able to easily pass in the current validator set in a way that manages the BLS public keys correctly.